design/src/rfcs/rfc0020_service_builder.md

```diff
@@ -0,0 +1,286 @@
+
+### Handlers
+
+An core concept in the current service builder is the `Handler` trait:
+
+```rust
+pub trait Handler<T, Input> {
+    async fn call(self, req: http::Request) -> http::Response;
+}
+```
+
+Its purpose is to provide an even interface over closures of the form `FnOnce(Input) -> impl Future<Output = Output>` and `FnOnce(Input, State) -> impl Future<Output = Output>`. It's this abstraction which allows the customers to supply both `async fn operation0(input: Operation0Input) -> Operation0Output` and `async fn operation0(input: Operation0Input, state: Extension<S>) -> Operation0Output` to the service builder.
+
+We generate `Handler` implementations for said closures in [ServerOperationHandlerGenerator.kt](https://github.com/awslabs/smithy-rs/blob/458eeb63b95e6e1e26de0858457adbc0b39cbe4e/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerOperationHandlerGenerator.kt), for `Operation0` these are:
+
+```rust
+impl<Fun, Fut> Handler<(), Operation0Input> for Fun
+where
+    Fun: FnOnce(crate::input::Operation0Input) -> Fut,
+    Fut: Future<Output = Operation0Output>,
+{
+    async fn call(self, request: http::Request) -> http::Response {
+        let input = /* Create `Operation0Input` from `request: http::Request` */;
+
+        // Use closure on the input
+        let output = self(input).await;
+
+        let response = /* Create `http::Response` from `output: Operation0Output` */
+        response
+    }
+}
+
+impl<Fun, Fut> Handler<Extension<S>, Operation0Input> for Fun
+where
+    Fun: FnOnce(Operation0Input, Extension<S>) -> Fut,
+    Fut: Future<Output = Operation0Output>,
+{
+    async fn call(self, request: http::Request) -> http::Response {
+        let input = /* Create `Operation0Input` from `request: http::Request` */;
+
+        // Use closure on the input and fetched extension data
+        let extension = Extension(request.extensions().get::<T>().clone());
+        let output = self(input, extension).await;
+
+        let response = /* Create `http::Response` from `output: Operation0Output` */
+        response
+    }
+}
+```
+
+Creating `Operation0Input` from a `http::Request` and `http::Response` from a `Operation0Output` involves the HTTP binding traits and protocol aware deserialization. The structure [RuntimeError](https://github.com/awslabs/smithy-rs/blob/458eeb63b95e6e1e26de0858457adbc0b39cbe4e/rust-runtime/aws-smithy-http-server/src/runtime_error.rs#L53-L5) enumerates internal error cases such as serialization/deserialization failures, `extensions().get::<T>()` failures, etc. We omit error handling in the snippets above, but, in full, they also involve protocol aware conversions from the `RuntimeError` and `http::Response`. The reader should make note of the influence of the model/protocol on the different sections of this procedure.
```

Comment:
```suggestion
Creating `Operation0Input` from a `http::Request` and `http::Response` from a `Operation0Output` involves the [HTTP binding traits](https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html) and protocol aware deserialization. The structure [RuntimeError](https://github.com/awslabs/smithy-rs/blob/458eeb63b95e6e1e26de0858457adbc0b39cbe4e/rust-runtime/aws-smithy-http-server/src/runtime_error.rs#L53-L5) enumerates error cases such as serialization/deserialization failures, `extensions().get::<T>()` failures, etc. We omit error handling in the snippets above, but, in full, they also involve protocol aware conversions from the `RuntimeError` and `http::Response`. The reader should make note of the influence of the model/protocol on the different sections of this procedure.
```

Despite not being `pub`, `RuntimeError` is not internal in that it gets exposed to the clients over the wire.

Reply:
I've worded this too roughly. What I'm trying to get at is that there are protocol specific failures which occur outside of the handler which can get captured by `RuntimeError`. What I'm going to argue for later is that, perhaps if we are going to have protocol specific routers, then perhaps it makes sense to have protocol specific errors (which all enjoy something akin to a `IntoResponse` trait).